### DISCUSSION TITLE
Help Imagemagick Standalone Mac not working

### DISCUSSION THREAD

**Discussion by bioworldmerch** [ImageMagick/ImageMagick#5215] [Help]
Im trying to install imagemagick as a standalone binary for Mac but been unsuccessful. 
I've racked my brain for days and searched many places and need some advice on installing a standalone Imagemagick without any dependencies. I believe the only edits I needed in Brew were:

`"--enable-delegate-build",
  "--without-frozenpaths",
"--disable-dependency-tracking",`

And get the following error:
`dyld: Library not loaded: /usr/local/opt/libomp/lib/libomp.dylib`

I'm using brew edit and the following:

` args = [
      "--with-quantum-depth=32",
      "--enable-osx-universal-binary=no",
      "--enable-delegate-build",
      "--without-frozenpaths",
      "--prefix=#{prefix}",
      "--disable-dependency-tracking",
      "--disable-silent-rules",
      "--disable-opencl",
      "--disable-shared",
      "--enable-static",
      "--with-freetype=yes",
      "--with-gvc=no",
      "--with-modules",
      "--with-openjp2",
      "--with-openexr",
      "--with-webp=yes",
      "--with-heic=yes",
      "--with-raw=yes",
      "--with-gslib",
      "--with-gs-font-dir=#{HOMEBREW_PREFIX}/share/ghostscript/fonts",
      "--with-lqr",
      "--without-fftw",
      "--without-pango",
      "--without-wmf",
      "--enable-openmp",`

If I try changing to ` "--enable-osx-universal-binary=yes",` The install fails totally. I get the following:
`Last 15 lines from /Users/myname/Library/Logs/Homebrew/imagemagick/02.make:
./MagickCore/utility-private.h:327:5: error: variable has incomplete type 'struct utimbuf'
    timestamp;
    ^
./MagickCore/utility-private.h:326:10: note: forward declaration of 'struct utimbuf'
  struct utimbuf
         ^
libtool: compile: mv -f "animate.o" "MagickWand/libMagickWand_7_Q32HDRI_la-animate.o"
libtool: compile: mv -f "display.o" "MagickWand/libMagickWand_7_Q32HDRI_la-display.o"
libtool: compile: mv -f "deprecate.o" "MagickWand/libMagickWand_7_Q32HDRI_la-deprecate.o"
1 error generated.
make[1]: *** [MagickWand/libMagickWand_7_Q32HDRI_la-convert.lo] Error 1
libtool: compile: mv -f "conjure.o" "MagickWand/libMagickWand_7_Q32HDRI_la-conjure.o"
libtool: compile: mv -f "compare.o" "MagickWand/libMagickWand_7_Q32HDRI_la-compare.o"
libtool: compile: mv -f "composite.o" "MagickWand/libMagickWand_7_Q32HDRI_la-composite.o"
make: *** [install] Error 2

Do not report this issue to Homebrew/brew or Homebrew/core!
`

I also tried removing the following lines:
  `on_macos do
    depends_on "libomp"
  end`

The install ran and made mention of installing libomp but still get same error.

I would be grateful for any advice!

**Comment by bioworldmerch** [comment]
I may have it figured out but have to test on a few Macs..... I followed Anthony'y directions for building personal copy of Imagemagick from the following old url https://www.imagemagick.org/Usage/api/#personal but modified the ./configure to the following:

`./configure --prefix=$MAGICK_HOME --with-quantum-depth=32 --with-xml --disable-installed --enable-delegate-build --disable-shared --enable-static --with-modules=no --enable-hdri=no --with-heic=no --without-magick-plus-plus --with-quantum-depth=32 --without-frozenpaths   --disable-silent-rules --disable-dependency-tracking --enable-zero-configuration --without-gslib --with-png=yes --with-jpeg=yes --with-jp2=yes --with-freetype=yes  --disable-osx-universal-binary --without-x --without-lcms --without-pango --without-bzlib --without-webp --without-perl --without-gvc --without-zip`

I'm concerned in the configure.xml file about 2 places:

` <configure name="CFLAGS" value="-I/opt/local/include/libxml2 -I/opt/local/include/libpng16 -I/opt/local/include -I/opt/local/include -I/opt/local/include -I/opt/local/include/ossp -I/opt/local/include/freetype2 -I/opt/local/include/libpng16 -I/opt/local/include  -I/opt/local/include  -Wall -g -O2 -mtune=core2 -fexceptions -D_FORTIFY_SOURCE=0 -pthread -DMAGICKCORE_HDRI_ENABLE=0 -DMAGICKCORE_QUANTUM_DEPTH=32"/>`

and 

` <configure name="LIBS" value="   -ltiff   -ljpeg    -L/opt/local/lib -lpng16      -L/opt/local/lib -lfontconfig -lfreetype          -L/opt/local/lib -llzma       -L/opt/local/lib -lxml2  -L/opt/local/lib -lz     -lm    -lpthread  "/>`

if I change those to point at a folder within my portable imagemagick folder to something like -L/Users/me/apps/im/lib/libname will Magick binary find them ok? Can I use relative paths? Not sure how.

Thank you for any of your advice! Once figured out Ill post the complete walkthrough for others to use on Mac in the future.

**Comment by snibgo** [comment]
Those aren't my directions. The bottom of the page says the author was Anthony Thyssen.

I have no experience with configure.xml, or building on a Mac. Sorry.

**Reply by bioworldmerch** [reply]
Sorry...edited above

**Comment by bioworldmerch** [comment]
Ok so I got Static standalone Imagemagick working on Mac...finally at least for me works on 3 different Macs moving the binary around. If anyone has some suggestions or changes feel free to insert your feedback. No guarantees it will work for you...Heres the steps I followed: 

1. Download the latest version of Imagemagick release or latest from Github repo or use curl command. `https://download.imagemagick.org/ImageMagick/download/ImageMagick-7.1.0-37.tar.xz`
2. `$ export MAGICK_HOME=$HOME/Imagemagick`
3. `$ tar xvzf ImageMagick-7.1.0-37.tar.gz`
4. `$ cd ImageMagick-7.1.0-37`
5. Download any delegates you want from `https://download.imagemagick.org/ImageMagick/download/delegates/`
6. Move the expanded folders into the ImageMagick-7.1.0-37 folder if you haven't downloaded them directly into the folder by using curl.
7. simplify folder names into jpeg/lbpng etc for the delegates for easier navigation
8. CD into each delegate folder you want to install. ie `$ cd JPEG`
9. `$ ./configure --enable-static --disable-shared --prefix=$MAGICK_HOME/lib/insert libName here like libPNG`
10. `$ cd ..` (up one into Imagemagick)
11. Repeat for each delegate folder and  do same configure command replacing the libName for example JPEG or Freetype
12. When completed configuring delegates `$ cd ..` into imagemagick folder
*** If your on Mac and wondering why JPEG download errors on configure make sure you download the https://download.imagemagick.org/ImageMagick/download/delegates/jpegsrc.v9b.tar.gz file and not the zip file. The zip will not install on Mac its specific to windows for some reason. I never would have guessed this would be the case.
13. `$ rm config.status config.log`
14. Figure out what delegates/config you want. Mine was: `$ ./configure --prefix=$MAGICK_HOME --with-quantum-depth=32 --with-xml --disable-installed --enable-delegate-build --disable-shared --enable-static --with-modules=no --enable-hdri=no --with-heic=no --without-magick-plus-plus --with-quantum-depth=32 --without-frozenpaths   --disable-silent-rules --disable-dependency-tracking --enable-zero-configuration --without-gslib --with-png=yes --with-jpeg=no --without-jp2 --disable-osx-universal-binary --without-x --without-lcms --without-freetype --without-pango --without-bzlib --without-webp --without-perl --without-gvc --without-zip --with-fontconfig=no` for more details see: https://imagemagick.org/script/advanced-linux-installation.php
15. `$ make clean`
16. `$ make` 
17. `$ make install`
18. `$ export PATH="$MAGICK_HOME/bin:$PATH"` 
19. `$ export LD_LIBRARY_PATH="$MAGICK_HOME/lib:$LD_LIBRARY_PATH"`
20. `$ exec "$@"`
21. Test your install using `magick --version` or `magick  -list configure` if no errors success. Move the binaries wherever you want to test on another computer. I'm 90% you can remove the --disable-installed if you don't want all the extra folders. I was playing it safe in case the libraries were missing and copied them into the lib folder in Imagemagick just in case.

I'm sure these steps could be simplified or fine tuned. If others confirm it worked for them, it will be helpful. Apologies in advance if it doesn't work for you. I came close to doing with Brew and using Brew Edits but the delegates were not all available. Hope this helps someone else.

**Reply by fmw42** [reply]
You have enabled 32-bit integer compile (Q32 without HDRI). Is that what you wanted? If you need 32-bit float, then you will need to install with HDRI enabled.

**Reply by bioworldmerch** [reply]
Thanks Ill add that. I was just using the 32 bit for the define connected-components to gain a higher max number. Do I still need HDRI?

**Reply by fmw42** [reply]
No, probably not.

**Reply by bioworldmerch** [reply]
Thanks as always!

**Reply by ivanooi** [reply]
do you need to place the binaries in a specific folder? thanks and looks promissiong! :-D

**Reply by bioworldmerch** [reply]
not that I have found so far. I added --disable-installed because it builds a nice structure and includes all the extras as an organized folders but I moved the Binaries inside BIN and they work without any supporting files. If you don't want it installed in custom location and to exclude the extra stuff just remove the --disable-installed

**Reply by speedgeil** [reply]
Thanks for posting this. I have no problem compiling the standalone Imagemagick. For me it's the delegates which are causing the problems. I followed your directions above but i'm getting errors (test failed) for tiff and jpeg after running configure. I guess i'm missing further dependencies to get this running. But i don't know what dependencies.
Also the docs are not clear on how to name the delegate folders in the imagemagick folder. 

For example when naming the delegate folder libjpeg, the configure script won't find the jpeg delegate. But it will find it when naming it jpeg. Same for tiff. So i guess the naming must follow the later file exension the delegates are supposed to work with. But what about libraw? It works with several file extensions like dng, cr2, nef.... How would i name it? Or lcms? Opening the configure script in a text editor gives some hints. But it will still not work because of missing dependencies, that are not documented.

Whatever, i did not manage to compile a standalone Imagemagick with one single delegate. Although i pulled the latest versions from the web. This is a complete trial-error-PITA!

**Reply by tim-sutherland** [reply]
Anyone have any luck recently?  I am not having any success compiling with delegates either.  I can get it to build but when I configure it either says no for all the delegates or it says no(test failed).

**Comment by emendelson** [comment]
@bioworldmerch - Not quite a standalone binary, but it gets the job done:

https://stackoverflow.com/questions/77767021/how-to-build-a-portable-version-of-imagemagick-for-macos-monterey-ventura-sonoma/77767022#77767022

**Comment by leleliu008** [comment]
look this: https://github.com/ImageMagick/ImageMagick/discussions/7054#discussioncomment-9571533
